libpromises/evalfunction.c

```diff
@@ -6659,6 +6659,158 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
     return (FnCallResult) { FNCALL_SUCCESS, (Rval) { json, RVAL_TYPE_CONTAINER } };
 }
 
+static int ContainerComparator(JsonElement const *tits,
+                               JsonElement const *ass,
+                               void *user_data)
+{
+    size_t *index = (size_t *)user_data;
+
+    char const *left = JsonPrimitiveGetAsString(JsonAt(tits, *index));
+    char const *right = JsonPrimitiveGetAsString(JsonAt(ass, *index));
+    return strcmp(left, right);
+}
+
+static FnCallResult FnCallClassExprFilter(EvalContext *ctx,
+                                          ARG_UNUSED Policy const *policy,
+                                          FnCall const *fp,
+                                          Rlist const *args)
+{
+    if (args == NULL || args->next == NULL || args->next->next == NULL)
+    {
+        FatalError(ctx, "Function %s requires at least 3 arguments",
+                   fp->name);
+    }
+
+    char const *path = RlistScalarValue(args);
+    args = args->next;
+    size_t const col = IntFromString(RlistScalarValue(args));
+    args = args->next;
+    char const *delim = RlistScalarValue(args);
+    args = args->next;
+    bool const has_heading = BooleanFromString(RlistScalarValue(args));
+    args = args->next;
+    size_t sort_col = IntFromString(RlistScalarValue(args));
+
+    FILE *csv_file = safe_fopen(path, "r");
+    if (csv_file == NULL)
+    {
+        FatalError(ctx, "Failed to read file %s: %s",
+                   path, GetErrorStrFromCode(errno));
+    }
+
+    Seq *heading_seq = NULL;
+    JsonElement *json = JsonArrayCreate(50);
+    char *line;
+    size_t line_len = 0;
+
+    // Current line number, for debugging
+    size_t line_number = 0;
+
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
+    {
+        Seq *seq = SeqParseCsvString(line);
+        ++line_number;
+        free(line);
+        if (seq == NULL)
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Failed to parse line %lu, line ignored.",
+                fp->name, line_number);
+            continue;
+        }
+
+        if (line_len == 0)
+        {
+            line_len = SeqLength(seq);
```

Comment:
What if `SeqLength` is 0? Looking at the implementation of `SeqParseCsvString()` it seems like it cannot return empty sequence. I would add an assert here `assert(line_len > 0)`.

Reply:
If you want to skip empty lines, I think those are returned as a sequence of 1 empty string.